### ISSUE TITLE
Performance parity with clBLAS (for small sizes)

### ISSUE THREAD

**Issue by gcp** [CNugteren/CLBlast#82]
Although complete performance parity with clBLAS is likely out of scope for this project, because clBLAS contains very specific kernels to work around limitations of AMD devices, we'd preferably still be as close as possible to its performance.

Of particular interest are situations where clBLAS outperforms CLBLast on NVIDIA or Intel hardware.

As described in issue #73, small or irregular configurations like (128,361,1152) are up to a factor of 3x slower on AMD, and 2x on NVIDIA, compared to clBLAS.

**Comment by CNugteren** [comment]
OK, like we discussed I'll work on a direct version of GEMM that doesn't require any pre and post-processing kernels in a new branch (`gemm_direct`).

For reference, this is the current performance on an AMD R9 M370X:
```
./clblast_client_xgemm -device 2 -m 128 -n 361 -k 1152 -warm_up
                                                                                             | <--       CLBlast       --> | <--       clBLAS        --> |
        m;        n;        k;   layout;   transA;   transB;      lda;      ldb;      ldc;       ms_1; GFLOPS_1;    GBs_1;     ms_2; GFLOPS_2;    GBs_2;  
      128;      361;     1152;      101;      111;      111;     1152;      361;      361;       2.10;     50.7;      1.2;     0.90;    117.9;      2.9;  
```
Indeed, a lot slower than clBLAS. And a lot slower than for larger sizes on the same device (~500 GFLOPS). I'll keep you updated.

**Comment by CNugteren** [comment]
I have a first working but very slow version in place (in `gemm_direct` branch). There are quite a couple of things I had/have to handle, here a summary:

1. Support for arbitrary values of `m` and `n`. This is implemented using a big if-else statement in the kernel. The else part is only executed on the incomplete border parts in `m` and `n` dimensions, so it doesn't matter too much that this part is slow and has extra bound-checks.
2. Support for arbitrary values of `k`. This is implemented by running the inner k-loop not to the very end, but only to complete tiles in the `k` dimension. The remainder is carried out in a not-unrolled non-tiling fashion, but that should be OK because this should be only a small part of the total work.
3. Support for complex conjugation of A and B. This is implemented with if-statements that are optimised away by the compiler if complex values are not used, so that should be OK for regular SGEMM and DGEMM.
4. Support for non-zero offset values. This is easily handled with almost no performance impact in case the offsets are actually zero. However, vector loads can only be used if these offset values are multiples of the vector widths, so that has to be implemented properly and with an extra if-statement.
6. Support for transposing A, B, and C. This is easy to do in the remainder parts (points 1 and 2), but not easy to do when performance matters. One issue is coalesced memory accesses and another is vector loads and stores. The loading part could perhaps be solved by caching squared (rectangular not allowed I think) blocks into local memory. The storing part could probably be handled as well if output blocks are squared and sufficiently large (e.g. 8 by 8?).
6. Support for arbitrary values of `lda`, `ldb`, and `ldc`. This can be tricky in combination with vectorisation when `ldX % vector_width != 0`. Not sure how to handle this yet.

In summary, this is not trivial, there is still some work to do. I'll probably try to reverse engineer some parts of clBLAS, see how they handle these issues.

**Comment by CNugteren** [comment]
Making some progress on the direct version of the GEMM kernel for smaller inputs (see `gemm_direct` branch). Below are some initial performance comparisons between the 'old' version (indirect kernel with pre and post-processing kernels) and the 'new' direct kernel tuned for small sizes.

On AMD R9 M370X:
```
./clblast_client_xgemm -m 256 -n 256 -k 1152 -warm_up -runs 100
                                                                                             | <--       CLBlast       --> | <--       clBLAS        --> |
        m;        n;        k;   layout;   transA;   transB;      lda;      ldb;      ldc;            ms_1; GFLOPS_1;              ms_2; GFLOPS_2;
      256;      256;     1152;      101;      111;      111;     1152;      256;      256;            1.14;    132.8;              1.08;    140.2;       <--- old
      256;      256;     1152;      101;      111;      111;     1152;      256;      256;            0.71;    213.6;              1.20;    125.7;       <--- new direct kernel

./clblast_client_xgemm -m 128 -n 361 -k 1152 -warm_up -runs 100
                                                                                             | <--       CLBlast       --> | <--       clBLAS        --> |
        m;        n;        k;   layout;   transA;   transB;      lda;      ldb;      ldc;            ms_1; GFLOPS_1;              ms_2; GFLOPS_2;
      128;      361;     1152;      101;      111;      111;     1152;      361;      361;            1.30;     82.0;              0.82;    129.8;       <--- old
      128;      361;     1152;      101;      111;      111;     1152;      361;      361;            2.20;     48.5;              0.92;    116.3;       <--- new direct kernel
```

For the `-m 256 -n 256 -k 1152` configuration performance was actually already on par with clBLAS and is now significantly better. For the `-m 128 -n 361 -k 1152` case performance was already quite OK (did something change or was the previous measurement incorrect?), but got a lot worse with the new direct kernel. Still some work to do to get that faster as well!

**Comment by CNugteren** [comment]
Fixed that performance issue as well. Performance on AMD R9 M370X:
```
./clblast_client_xgemm -m 256 -n 256 -k 1152 -warm_up -runs 100
                                                                                             | <--       CLBlast       --> | <--       clBLAS        --> |
        m;        n;        k;   layout;   transA;   transB;      lda;      ldb;      ldc;            ms_1; GFLOPS_1;              ms_2; GFLOPS_2;
      256;      256;     1152;      101;      111;      111;     1152;      256;      256;            1.14;    132.8;              1.08;    140.2;       <--- old
      256;      256;     1152;      101;      111;      111;     1152;      256;      256;            0.73;    207.1;              1.08;    140.1;       <--- new direct kernel

./clblast_client_xgemm -m 128 -n 361 -k 1152 -warm_up -runs 100
                                                                                             | <--       CLBlast       --> | <--       clBLAS        --> |
        m;        n;        k;   layout;   transA;   transB;      lda;      ldb;      ldc;            ms_1; GFLOPS_1;              ms_2; GFLOPS_2;
      128;      361;     1152;      101;      111;      111;     1152;      361;      361;            1.30;     82.0;              0.82;    129.8;       <--- old
      128;      361;     1152;      101;      111;      111;     1152;      361;      361;            1.05;    101.5;              0.74;    144.1;       <--- new direct kernel
```

Timing for these short kernels is tricky, and so is tuning for small sizes. I'll test on some other devices to verify that the new direct kernel is correct & fast. Then I'll merge it into the `development` branch and it's open for users to tune for their devices!

**Comment by CNugteren** [comment]
I've done some proper benchmarking now. Attached are two graphs ('new' with the new direct kernel for `m*n*k < 512*512*512` and 'old' with only the indirect kernel) for the Radeon M370X:
[SGEMM_M370X_new.pdf](https://github.com/CNugteren/CLBlast/files/517908/SGEMM_M370X_new.pdf)
[SGEMM_M370X_old.pdf](https://github.com/CNugteren/CLBlast/files/517907/SGEMM_M370X_old.pdf)

As you can see, performance is in general better than clBLAS with exceptions of some peaks. Performance also improved for the smallest sizes, although it is a bit tricky to see on this graph. Gains are not huge though. Also note that benchmarking is not easy for small sizes: these are already averages over 50 runs but still the clBLAS results differ a bit.

This is for an AMD GPU. I still have to test a bit on other devices before I merge it in.

**Comment by CNugteren** [comment]
Here are some other results for a Geforce GTX750 Ti and an Intel Iris Pro GPU:
[SGEMM_GTX750Ti_new.pdf](https://github.com/CNugteren/CLBlast/files/519345/SGEMM_GTX750Ti_new.pdf)
[SGEMM_GTX750Ti_old.pdf](https://github.com/CNugteren/CLBlast/files/519346/SGEMM_GTX750Ti_old.pdf)
[SGEMM_IrisPro_new.pdf](https://github.com/CNugteren/CLBlast/files/519347/SGEMM_IrisPro_new.pdf)
[SGEMM_IrisPro_old.pdf](https://github.com/CNugteren/CLBlast/files/519348/SGEMM_IrisPro_old.pdf)
(note that these comparisons are against an untuned clBLAS)

For the GTX750 Ti the advantage is clear, perhaps it would actually make sense to go beyond the `512*512*512` limit for the small kernel. For the Iris Pro the advantage is only in the very small sizes, so there perhaps this limit is too high. Of course this can be tweaked.

**Comment by CNugteren** [comment]
In the meantime all of this is in the `development` branch. Also, some of the included benchmarks in the CLBlast `doc/performance` have been updated (tested against a tuned clBLAS).

I am closing this issue now. Feel free to re-open it if you still think performance for small sizes is not sufficient.
